modin/experimental/batch/test/test_pipeline.py

```diff
@@ -0,0 +1,750 @@
+            return df
+
+        def reducer(dfs):
+            new_cols = "".join([str(df["new_col"].values[0]) for df in dfs])
+            dfs[0]["new_col1"] = new_cols
+            return dfs[0]
+
+        pipeline = PandasQueryPipeline(df, num_partitions=24)
+        pipeline.add_query(
+            new_col_adder,
+            fan_out=True,
+            reduce_fn=reducer,
+            pass_partition_id=True,
+            is_output=True,
+            output_id=20,
+        )
+        pipeline.add_query(
+            lambda df: pandas.concat([df] * 1000), repartition_after=True
+        )
+        pipeline.add_query(
+            lambda df: df.drop(columns=["new_col"]), is_output=True, output_id=21
+        )
+
+        def to_csv(df, partition_id):
+            df.to_csv(f"{partition_id}.csv")
+            return df
+
+        pipeline.add_query(to_csv, is_output=True, output_id=22, pass_partition_id=True)
+
+        def post_proc(df, o_id, partition_id):
+            df["new_col"] = f"{o_id} {partition_id}"
+            return df
+
+        new_dfs = pipeline.compute_batch(
+            postprocessor=post_proc,
+            pass_partition_id=True,
+            pass_output_id=True,
+            final_result_func=lambda df: df.iloc[-1],
+        )
+        correct_df = pd.DataFrame([[0, 1, 2]])
+        correct_df["new_col"] = "20 0"
+        correct_df["new_col1"] = "".join([str(i) for i in range(24)])
+        df_equals(correct_df.iloc[-1], new_dfs[20])
+        correct_df = pd.concat([correct_df] * 1000)
+        correct_df = correct_df.drop(columns=["new_col"])
+        correct_df["new_col"] = "21 0"
+        df_equals(correct_df.iloc[0], new_dfs[21])
+        for i in range(24):
+            assert exists(
+                f"{i}.csv"
+            ), "CSV File for Partition {i} does not exist, even though dataframe should have been repartitioned."
+            remove(f"{i}.csv")
+        assert 22 in new_dfs, "Output for output id 22 not generated."
+
+
+@pytest.mark.skipif(
+    Engine.get() == "Ray",
+    reason="Ray supports the Batch Pipeline API",
+)
+def test_pipeline_unsupported_engine():
+    """
+    This test ensures that trying to use the Pipeline API with an unsupported Engine raises errors.
+    """
+    # Check that pipeline does not allow `Engine` to not be Ray.
+    df = pd.DataFrame([[1, 2, 3], [4, 5, 6]])
+    with pytest.raises(
+        NotImplementedError,
+        match="Batch Pipeline API is only implemented for Ray Engine.",
+    ):
+        PandasQueryPipeline(df)
+
+    eng = Engine.get()
+    Engine.put("Ray")
+    # Check that even if Engine is Ray, if the df is not backed by Ray, the Pipeline does not allow initialization.
+    with pytest.raises(
+        NotImplementedError,
+        match="Batch Pipeline API is only implemented for Ray Engine.",
+    ):
+        PandasQueryPipeline(df, 0)
+    new_df = pd.DataFrame([[1, 2, 3], [5, 6, 7]])
+    pipeline = PandasQueryPipeline(new_df)
```

Comment:
```suggestion
    ray_df = pd.DataFrame([[2]])
    pipeline = PandasQueryPipeline(ray_df)
```

Reply:
I think `ray_df` is not relevant here. `new_df` is okay.

Reply:
I think this  makes sense - I've changed it to `ray_df`!

Reply:
Why do you think this makes sense? It might be confusing for a newbie of Modin, which may think of an object from Ray itself. It is a Modin dataframe because it is created via modin.pandas.DataFrame.

Reply:
`new_df` doesn't tell us anything about why we can make a pipeline out of this df but not the previous one. `ray_df` seems to make it clear that the df is backed by ray, as opposed to the previous `df`. There are so few lines between where this object is created and the last time it's used that I don't expect anyone to think it's really a ray object. `ray_backed_df` seems like a helpful name that could be a little clearer.

Reply:
When the user works with a Modin object (dataframe, pipeline, etc.), all internals should be hided from the user and all the work should be transparent for him. He hasn't to care of which engine the Dataframe is backed. Moreover, we will likely have a pipeline object implemented with other engines.

Reply:
I agree that in general a user should not have to think about the internals of the Modin dataframe, including which engine is backing the dataframe. This unit test, though, is not primarily for users. It's primarily for people maintaining Modin, and the code will probably be read next by a Modin developer either trying to figure out why it's broken or modifying it after adding support for engines. This part of the test is specifically checking what happens when someone tries to create a pipeline from a dataframe backed by Ray storage, as opposed to one not backed by Ray storage, all while the engine is Ray: the former succeeds whereas the latter (tested in the previous few lines) fails. So I still think `ray_df` or, better, `ray_backed_df` is the right term.

Reply:
I'll rename to `df_on_ray_engine` to make it explicit!

Reply:
ok